recipes/openblas/all/conanfile.py

```diff
@@ -19,134 +23,212 @@ class OpenblasConan(ConanFile):
+                    return "gomp"
+            if self.settings.compiler == "clang":
+                if Version(self.settings.compiler.version).major > "8":
+                    return "omp"
+        return None
 
     @property
-    def _build_subfolder(self):
-        return "build_subfolder"
+    def _fortran_compiler(self):
+        comp_exe = self.conf.get("tools.build:compiler_executables")
+        if comp_exe and 'fortran' in comp_exe:
+            return comp_exe["fortran"]
+        return environ.get('FC')
 
     def export_sources(self):
-        self.copy("CMakeLists.txt")
+        export_conandata_patches(self)
 
     def config_options(self):
         if self.settings.os == "Windows":
             del self.options.fPIC
 
+        # Build LAPACK by default if possible w/o Fortran compiler
+        if Version(self.version) >= "0.3.21":
+            self.options.build_lapack = True
+
     def configure(self):
         if self.options.shared:
-            del self.options.fPIC
+            self.options.rm_safe("fPIC")
+
+    def layout(self):
+        cmake_layout(self, src_folder="src")
+
+    def package_id(self):
+        if self.info.options.build_lapack:
+            # Capture fortran compiler in package id (if found)
+            f_compiler = self._fortran_compiler
+            if not f_compiler:
+                self.output.warning("None or unknown fortran compiler was used.")
+                f_compiler = True
+            self.info.options.build_lapack = f_compiler
 
     def validate(self):
-        if hasattr(self, "settings_build") and tools.cross_building(self, skip_x64_x86=True):
+        if cross_building(self, skip_x64_x86=True):
             raise ConanInvalidConfiguration("Cross-building not implemented")
 
+        if self.options.use_thread and self.options.use_openmp:
+            raise ConanInvalidConfiguration("Both 'use_thread=True' and 'use_openmp=True' are not allowed")
+
+        if self.settings.os == "Windows" and self.options.build_lapack and self.options.use_openmp:
+            # In OpenBLAS cmake/system.cmake: Disable -fopenmp for LAPACK Fortran codes on Windows
+            self.output.warning("OpenMP is disabled on LAPACK targets on Windows")
+
+    def build_requirements(self):
+        if self.options.build_lapack and self.settings.os == "Windows":
+            self.tool_requires("ninja/1.11.1")
+
     def source(self):
-        tools.get(
-            **self.conan_data["sources"][self.version],
-            strip_root=True,
-            destination=self._source_subfolder
-        )
+        get(self, **self.conan_data["sources"][self.version], strip_root=True)
 
-    @functools.lru_cache(1)
-    def _configure_cmake(self):
-        cmake = CMake(self)
+    def generate(self):
 
+        ninja_generator = "Ninja" if self.settings.os == "Windows" and \
+            self.options.build_lapack else None
+        tc = CMakeToolchain(self, generator=ninja_generator)
```

Comment:
Is there a particular requirement to use ninja here? it's not common in cci recipes to be so explicit - ideally the user can use whatever their preference is.

Reply:
@joakimono that doesn't seem strictly true - [this section of the wiki](https://github.com/xianyi/OpenBLAS/wiki/How-to-use-OpenBLAS-in-Microsoft-Visual-Studio#2-use-cmake-to-generate-visual-studio-solution-files) seems to suggest that it works. Perhaps it's just a matter of fleshing out the `validate()` section to make sure it's built with the right combination of configuration when the generator is visual studio? You should be able to check this with something like

```python
if "Visual Studio" in self.conf.get("tools.cmake.cmaketoolchain:generator") and self.options.get_safe("dynamic_arch"):
    # raise exception
```
Alternatively, I think something preferable to specifying to use Ninja all the time would be to override this in `configure` - though that's still questionable... (I'm not sure if the `conf` values are considered read only) something like
```python
def configure(self):
    generator = self.conf.get("tools.cmake.cmaketoolchain:generator")
    if "Visual Studio" in generator:
        self.output.warning(f"{generator} is not compatible with {self.name}. Updating tools.cmake.cmaketoolchain:generator to use 'Ninja' instead.")
        self.conf.update("tools.cmake.cmaketoolchain:generator", "Ninja")
```

Regardless, I don't like that this forces the use of ninja - especially where any of the other generators may be equally valid except for the visual studio one.

Reply:
My current proposal only force to use the Ninja generator in cases I have confirmed by trial that Visual Studio's standard generator fails to successfully build the project. It should be noted that the default value of the `generator` argument of `CMakeToolchain` is `None`, for which `ninja_generator` in the snippet above also evaluates when Ninja is not needed. To be clear; Ninja is only needed when building on Windows with option `build_lapack=True`. The only exception, if i recall correctly, is if it is built without a Fortran compiler. I could add another check to not force Ninja if no Fortran compiler is used.

I will make some tests regarding your suggestions above, especially regarding get/set of `tools.cmake.cmaketoolchain:generator`. However, I'm not sure if this variable is defined by default, which makes checking for 'Visual Studio' pointless unless the user explicitly has specified a generator. On second thought, it might be useful to inform the user that their requested generator has been overridden.

Reply:
yes, you're right - it's not defined by default, and it will just choose the platform default. On second look, I think what you've got is probably fine for most cases, but we should try to adhere to the law of least surprise. In that vein, I think you should do the following:
1. Add a warning output to indicate that the `Ninja` generator is being used when the selection logic is satisfied
2. Check that `tools.cmake.cmaketoolchain:generator` hasn't been set to something other than `Visual Studio 20xx` - if it has, you should prefer the user specified value instead. This could be something like

```python
conf_generator = self.conf.get("tools.cmake.cmaketoolchain:generator")
ninja_conditions = self.settings.os == "Windows" and self.options.build_lapack and (not conf_generator or "Visual Studio" in conf_generator):
generator = "Ninja" if ninja_conditions else None
if ninja_conditions:
    self.output.warning("The Visual Studio generator is not compatible with OpenBLAS when building LAPACK. Overwriting generator to 'Ninja'")
tc = CMakeToolchain(self, generator=generator)
```

Reply:
I have applied these suggested changes in the latest revision.